internal/users/db/users.go

```diff
@@ -0,0 +1,220 @@
+type UserDB struct {
+	Name  string
+	UID   uint32
+	GID   uint32
+	Gecos string // Gecos is an optional field. It can be empty.
+	Home  string
+	Shell string
+
+	// BrokerID specifies the broker the user last successfully authenticated with.
+	BrokerID string `yaml:"broker_id,omitempty"`
+}
+
+// userRow is the struct that is stored in the database.
+//
+// It prevents leaking of lastLogin, which is only relevant to the database.
+// TODO: The only consumer of this package is the users manager, which converts the UserDB into a types.UserEntry anyway,
+//
+//	so there is no need to hide the lastLogin field (which complicates the code).
+type userRow struct {
+	UserDB `yaml:",inline"`
+	// TODO: Why do we store the last login time in the database? It's not used anywhere.
+	LastLogin time.Time `yaml:"last_login,omitempty"`
+}
+
+type nullableUserRow struct {
+	UserDB
+	LastLogin sql.NullTime
+}
+
+// NewUserDB creates a new UserDB.
+func NewUserDB(name string, uid, gid uint32, gecos, dir, shell string) UserDB {
+	return UserDB{
+		Name:  name,
+		UID:   uid,
+		GID:   gid,
+		Gecos: gecos,
+		Home:  dir,
+		Shell: shell,
+	}
+}
+
+// UserByID returns a user matching this uid or an error if the database is corrupted or no entry was found.
+func (c *Database) UserByID(uid uint32) (UserDB, error) {
+	return userByID(c.db, uid)
+}
+
+func userByID(db queryable, uid uint32) (UserDB, error) {
+	query := fmt.Sprintf(`SELECT %s FROM users WHERE uid = ?`, publicUserColumns)
+	row := db.QueryRow(query, uid)
+
+	var u UserDB
+	err := row.Scan(&u.Name, &u.UID, &u.GID, &u.Gecos, &u.Home, &u.Shell, &u.BrokerID)
+	if errors.Is(err, sql.ErrNoRows) {
+		return UserDB{}, NoDataFoundError{key: strconv.FormatUint(uint64(uid), 10), table: "users"}
+	}
+	if err != nil {
+		return UserDB{}, fmt.Errorf("query error: %w", err)
+	}
+
+	return u, nil
+}
+
+// UserByName returns a user matching this name or an error if the database is corrupted or no entry was found.
+func (c *Database) UserByName(name string) (UserDB, error) {
+	query := fmt.Sprintf(`SELECT %s FROM users WHERE name = ?`, publicUserColumns)
+	row := c.db.QueryRow(query, name)
+
+	var u UserDB
+	err := row.Scan(&u.Name, &u.UID, &u.GID, &u.Gecos, &u.Home, &u.Shell, &u.BrokerID)
+	if errors.Is(err, sql.ErrNoRows) {
+		return UserDB{}, NoDataFoundError{key: name, table: "users"}
+	}
+	if err != nil {
+		return UserDB{}, fmt.Errorf("query error: %w", err)
+	}
+
+	return u, nil
+}
+
+// AllUsers returns all users or an error if the database is corrupted.
+func (c *Database) AllUsers() ([]UserDB, error) {
+	return allUsers(c.db)
+}
+
+func allUsers(db queryable) ([]UserDB, error) {
+	users, err := allUsersInternal(db)
+	if err != nil {
+		return nil, err
+	}
+
+	var res []UserDB
+	for _, u := range users {
+		res = append(res, u.UserDB)
+	}
+
+	return res, nil
+}
+
+func allUsersInternal(db queryable) ([]userRow, error) {
+	query := fmt.Sprintf(`SELECT %s FROM users`, allUserColumns)
+	rows, err := db.Query(query)
```

Comment:
As per what I've been [reading](https://github.com/mattn/go-sqlite3/issues/569), these calls shouldn't be happen concurrently, so maybe we should have a wrapper struct that manages the db locking via a mutex? So that we don't risk of accessing to it when it's locked? Or are we using another way to prevent this to happen?

Reply:
That's taken care of by the use of transactions. All functions which write to the database use transactions, which lock the database.